### PR TITLE
[FIX] base: detect errors during inheritance and use spec for expression

### PR DESCRIPTION
**Pull request by Gorash**
The inheritance for python expression (in modifiers) can use separators "and" and "or". The original expression will automatically be wrapped with parentheses to avoid any semantic change in the original expression.

task-3433721

linked merged pr: https://github.com/odoo/odoo/pull/104741

### CODE REVIEW

**Review by Feyensv — APPROVED** [review]

File: odoo/tools/template_inheritance.py [hunk 1]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
-                        separator = child.get('separator', ',')
-                        if separator == ' ':
-                            separator = None    # squash spaces
-                        to_add = (
-                            s for s in (s.strip() for s in child.get('add', '').split(separator))
-                            if s
+                    attribute = child.attrib.pop('name')
+                    value = None
+
+                    if child.get('add') or child.get('remove') or 'separator' in child.attrib:
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:211]: This condition is way too sloppy.  What happens if you only have attribute `separator`?  Nothing, no error.  There should be a test with only `separator`, which IMHO should raise an exception.
I suggest to be more specific: the element `<attribute>` should only have attributes:
- `name` (mandatory),
- either `add`, `remove` (both with optional `separator`) or none of them,
- any attribute that starts with `data-oe-`

File: odoo/tools/template_inheritance.py [hunk 2]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+                        if child.text:
+                            raise ValueError(
+                                _("Forbidden text %s") %
+                                repr(child.text)
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:215]: The API of `_()` now takes the `%s` parameters as positional parameters.
```suggestion
                                _("Element <attribute> with 'add' or 'remove' cannot contain text %s", repr(child.text))
```

File: odoo/tools/template_inheritance.py [hunk 3]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+
+                        if attribute in ('invisible', 'column_invisible', 'required', 'readonly', 't-if', 't-elif') or attribute.startswith('decorator-'):
+                            # python expression
+                            value = f' ({value}) ' if value else value
+                            if child.get('separator') not in (' and ', ' or '):
+                                raise ValueError(
+                                    _("Wrong separator %s for %s python expression please use only ' and ' or ' or ' separator") %
+                                    (repr(child.get('separator')), repr(attribute))
+                                )
+                            separator = child.attrib.pop('separator').strip()
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:227]: If you do trim the surrounding spaces, then there is no point in having `separator=" and "`.  Let's say that the only valid separators are `"and"` and `"or"`, and let the inheritance mechanism add spaces around them.

File: odoo/tools/template_inheritance.py [hunk 4]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+
+                            if child.get('remove'):
+                                remove = child.attrib.pop('remove')
+                                if remove == value:
+                                    value = ''
+                                elif remove in value:
+                                    is_constant = remove in ('True', 'False', '1', '0')
+                                    rf_remove = rf'(( +\()*)( *{re.escape(remove)} *)((\) +)*)' if is_constant else rf'(( +\()+)( *{re.escape(remove)} *)((\) +)+)'
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:235]: I honestly believe that the "remove" case of way too complicated.  The proof being that you've written a naive Python parser.  I suggest the following strategy, which any developer can understand: find the first occurrence of one of the strings `f" {separator} {remove}"`, `f"{remove} {separator} "` (pick the first one that matches) and remove it from the value.  If the removal must match parentheses, then the developer must put those parentheses in `remove`.  The least amount of magic is the better, IMHO.

File: odoo/tools/template_inheritance.py [hunk 5]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+                        if child.text:
+                            raise ValueError(
+                                _("Forbidden text %s") %
+                                repr(child.text)
+                            )
+                        value = node.get(attribute, '')
+
+                        if attribute in ('invisible', 'column_invisible', 'required', 'readonly', 't-if', 't-elif') or attribute.startswith('decorator-'):
+                            # python expression
+                            value = f' ({value}) ' if value else value
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:221]: This is making everything more complicated.  The add or remove operation is done on `value`, not `f" ({value}) "`.

File: odoo/tools/template_inheritance.py [hunk 6]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+                                            value = value.replace(f'{expression}{separator} ', ' ')
+                                        else:
+                                            break
+
+                                    if value == ' ':
+                                        value = ''
+
+                            if child.get('add'):
+                                add = child.attrib.pop('add')
+                                value = f"{value}{separator} ({add}) " if value else add
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:270]: If you follow my suggestions above, this becomes
```suggestion
                            if child.get('add'):
                                add = child.attrib.pop('add')
                                value = f"({value}) {separator} ({add})" if value else add
```
which will make the code simpler and the result more intuitive.

File: odoo/tools/template_inheritance.py [hunk 7]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+                            to_add = (
+                                s for s in (s.strip() for s in child.attrib.pop('add', '').split(separator))
+                                if s
+                            )
+                            to_remove = {s.strip() for s in child.attrib.pop('remove', '').split(separator)}
+                            values = (s.strip() for s in node.get(attribute, '').split(separator))
+                            value = (separator or ' ').join(itertools.chain(
+                                (v for v in values if v not in to_remove),
+                                to_add
+                            ))
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:287]: The existing code looks awfully complicated to me.  Why don't you make an ordered set with the `value`?
```suggestion
                            items = OrderedSet(s.strip() for s in node.get(attribute, '').split(separator))
                            for s in filter(None, child.attrib.pop('remove', '').split(separator)):
                                items.discard(s.strip())
                            for s in filter(None, child.attrib.pop('add', '').split(separator)):
                                items.add(s.strip())
                            value = (separator or ' ').join(items)
```

File: odoo/tools/template_inheritance.py [hunk 8]
```diff
@@ -205,23 +205,94 @@ def extract(spec):
...
+                    else:
+                        value = child.text or ''
+
+                    if any(not key.startswith('data-oe-') for key in child.attrib):
+                        raise ValueError(
+                            _("Invalid attribute %s") %
+                            ','.join(map(repr, child.attrib))
```
Comment (rco-odoo) [inline odoo/tools/template_inheritance.py:294]:
```suggestion
                    if invalids := [key for key in child.attrib if not key.startswith('data-oe-')]:
                        raise ValueError(
                            _("Invalid attributes %s in element <attribute>", ", ".join(map(repr, invalids)))
```

**Comment by rco-odoo** [comment]
@robodoo retry

**Pull request closed without merging**